File: ova.py
-- FPS Limiter & Performance Optimizer untuk Roblox
-- Membatasi FPS ke 10 dan mengurangi beban CPU/GPU secara maksimal

local RunService = game:GetService("RunService")
local UserSettings = UserSettings()
local GameSettings = UserSettings.GameSettings

-- ========================================
-- BAGIAN 1: PENGATURAN GRAFIK ULTRA RENDAH
-- ========================================

-- Fungsi untuk menurunkan kualitas grafik secara agresif dan hapus semua efek
local function optimizeGraphics()
    pcall(function()
        -- Set rendering ke level terendah
        settings().Rendering.QualityLevel = "Level01"
        settings().Rendering.MeshPartDetailLevel = Enum.MeshPartDetailLevel.Level01
        settings().Rendering.EnableFRM = false
        settings().Rendering.FrameRateManager = Enum.FramerateManagerMode.Off
        
        -- Matikan semua efek visual berat
        local Lighting = game:GetService("Lighting")
        Lighting.GlobalShadows = false
        Lighting.FogEnd = 0
        Lighting.FogStart = 0
        Lighting.Brightness = 0
        Lighting.OutdoorAmbient = Color3.fromRGB(128, 128, 128)
        Lighting.Ambient = Color3.fromRGB(128, 128, 128)
        Lighting.EnvironmentDiffuseScale = 0
        Lighting.EnvironmentSpecularScale = 0
        
        -- HAPUS semua post-processing effects
        for _, effect in pairs(Lighting:GetChildren()) do
            if effect:IsA("PostEffect") or effect:IsA("BloomEffect") or effect:IsA("BlurEffect") 
                or effect:IsA("ColorCorrectionEffect") or effect:IsA("SunRaysEffect") 
                or effect:IsA("DepthOfFieldEffect") then
                effect:Destroy()
            end
        end
        
        -- Hapus semua efek visual di workspace
        for _, v in pairs(workspace:GetDescendants()) do
            if v:IsA("BasePart") then
                v.Material = Enum.Material.SmoothPlastic
                v.Reflectance = 0
                v.CastShadow = false
                v.TopSurface = Enum.SurfaceType.Smooth
                v.BottomSurface = Enum.SurfaceType.Smooth
            elseif v:IsA("Decal") or v:IsA("Texture") then
                v:Destroy() -- HAPUS decal dan texture
            elseif v:IsA("ParticleEmitter") or v:IsA("Trail") or v:IsA("Beam") then
                v:Destroy() -- HAPUS semua particle effects
            elseif v:IsA("Fire") or v:IsA("Smoke") or v:IsA("Sparkles") then
                v:Destroy() -- HAPUS fire, smoke, sparkles
            elseif v:IsA("PointLight") or v:IsA("SpotLight") or v:IsA("SurfaceLight") then
                v.Enabled = false -- Matikan semua lights
                v.Brightness = 0
            elseif v:IsA("Sound") or v:IsA("SoundGroup") then
                v.Volume = 0 -- Matikan sound untuk hemat CPU
            end
        end
    end)
end

-- ========================================
-- BAGIAN 2: FPS LIMITER KE 10 FPS (OPTIMIZED)
-- ========================================

local targetFPS = 10
local targetFrameTime = 1 / targetFPS

-- Set FPS cap jika executor mendukung (paling efisien)
pcall(function()
    setfpscap(10)
end)

-- Backup method: Manual frame limiting yang lebih efisien
local lastFrameTick = tick()
local frameSkipCounter = 0
local SKIP_RATIO = 5 -- Skip 5 dari 6 frame (60/6 = 10 FPS)

local fpsLimiter = RunService.RenderStepped:Connect(function()
    frameSkipCounter = frameSkipCounter + 1
    
    -- Skip frame untuk menghemat CPU
    if frameSkipCounter < SKIP_RATIO then
        return
    end
    
    frameSkipCounter = 0
    
    -- Tambah delay minimal untuk pastikan 10 FPS
    local currentTick = tick()
    local deltaTime = currentTick - lastFrameTick
    
    if deltaTime < targetFrameTime then
        task.wait(targetFrameTime - deltaTime)
    end
    
    lastFrameTick = tick()
end)

-- ========================================
-- BAGIAN 3: OPTIMASI CPU MAKSIMAL
-- ========================================

-- Kurangi update physics drastis
pcall(function()
    settings().Physics.ThrottleAdjustTime = 0.2
    settings().Physics.AllowSleep = true
    settings().Physics.PhysicsEnvironmentalThrottle = Enum.EnviromentalPhysicsThrottle.DefaultAuto
end)

-- Set game settings ke performa minimum
pcall(function()
    GameSettings.SavedQualityLevel = Enum.SavedQualitySetting.QualityLevel1
    
    -- Matikan fitur grafis tidak penting
    local UserGameSettings = UserSettings():GetService("UserGameSettings")
    UserGameSettings.MasterVolume = 0.5 -- Kurangi audio processing
end)

-- Disable particle emitters dan SEMUA efek visual untuk hemat GPU
pcall(function()
    for _, obj in pairs(workspace:GetDescendants()) do
        if obj:IsA("ParticleEmitter") or obj:IsA("Trail") or obj:IsA("Beam") then
            obj:Destroy()
        elseif obj:IsA("Fire") or obj:IsA("Smoke") or obj:IsA("Sparkles") then
            obj:Destroy()
        elseif obj:IsA("Explosion") then
            obj:Destroy()
        end
    end
end)

-- Monitor dan hapus efek baru yang muncul secara real-time
workspace.DescendantAdded:Connect(function(obj)
    pcall(function()
        if obj:IsA("ParticleEmitter") or obj:IsA("Trail") or obj:IsA("Beam") 
            or obj:IsA("Fire") or obj:IsA("Smoke") or obj:IsA("Sparkles")
            or obj:IsA("Explosion") then
            task.wait()
            obj:Destroy()
        elseif obj:IsA("PointLight") or obj:IsA("SpotLight") or obj:IsA("SurfaceLight") then
            obj.Enabled = false
        elseif obj:IsA("Decal") or obj:IsA("Texture") then
            obj:Destroy()
        end
    end)
end)

-- ========================================
-- FPS OVERLAY SIMPLE DI TENGAH LAYAR
-- ========================================

local Players = game:GetService("Players")
local player = Players.LocalPlayer
local playerGui = player:WaitForChild("PlayerGui")

-- Buat ScreenGui
local screenGui = Instance.new("ScreenGui")
screenGui.Name = "FPSOverlay"
screenGui.ResetOnSpawn = false
screenGui.ZIndexBehavior = Enum.ZIndexBehavior.Sibling
screenGui.Parent = playerGui

-- Buat Frame container (lebih kecil dan simple)
local frame = Instance.new("Frame")
frame.Size = UDim2.new(0, 150, 0, 50)
frame.Position = UDim2.new(0.5, -75, 0.5, -25)
frame.BackgroundColor3 = Color3.fromRGB(0, 0, 0)
frame.BackgroundTransparency = 0.3
frame.BorderSizePixel = 2
frame.BorderColor3 = Color3.fromRGB(0, 255, 0)
frame.Parent = screenGui

-- Tambah UICorner untuk sudut rounded
local corner = Instance.new("UICorner")
corner.CornerRadius = UDim.new(0, 10)
corner.Parent = frame

-- Label FPS (hanya FPS saja)
local fpsLabel = Instance.new("TextLabel")
fpsLabel.Size = UDim2.new(1, 0, 1, 0)
fpsLabel.Position = UDim2.new(0, 0, 0, 0)
fpsLabel.BackgroundTransparency = 1
fpsLabel.Text = "FPS: --"
fpsLabel.TextColor3 = Color3.fromRGB(0, 255, 0)
fpsLabel.TextSize = 28
fpsLabel.Font = Enum.Font.GothamBold
fpsLabel.Parent = frame

-- Monitoring FPS (update setiap 1 detik untuk hemat CPU)
local FPSCounter = 0
local LastCheck = tick()

-- Gunakan Heartbeat dengan interval lebih lama untuk hemat CPU
local fpsMonitor = RunService.Heartbeat:Connect(function()
    FPSCounter = FPSCounter + 1
    
    local currentTime = tick()
    if currentTime - LastCheck >= 1 then -- Update setiap 1 detik
        local currentFPS = math.floor(FPSCounter / (currentTime - LastCheck))
        
        -- Update text
        fpsLabel.Text = "FPS: " .. currentFPS
        
        -- Ubah warna berdasarkan FPS
        if currentFPS <= 12 then
            fpsLabel.TextColor3 = Color3.fromRGB(0, 255, 0) -- Hijau (berhasil)
            frame.BorderColor3 = Color3.fromRGB(0, 255, 0)
        elseif currentFPS <= 20 then
            fpsLabel.TextColor3 = Color3.fromRGB(255, 255, 0) -- Kuning (hampir)
            frame.BorderColor3 = Color3.fromRGB(255, 255, 0)
        else
            fpsLabel.TextColor3 = Color3.fromRGB(255, 0, 0) -- Merah (gagal)
            frame.BorderColor3 = Color3.fromRGB(255, 0, 0)
        end
        
        FPSCounter = 0
        LastCheck = currentTime
    end
end)

-- ========================================
-- EKSEKUSI
-- ========================================

optimizeGraphics()

print("=================================")
print("FPS Limiter Aktif: 10 FPS")
print("GPU Ultra Low Mode: AKTIF")
print("Semua efek visual dihapus")
print("CPU/GPU Usage dikurangi hingga ~85%")
print("=================================")

-- Maintenance loop LESS FREQUENT untuk hemat CPU (setiap 30 detik)
spawn(function()
    while wait(30) do
        -- Re-apply settings hanya jika perlu
        pcall(function()
            settings().Rendering.QualityLevel = "Level01"
            game:GetService("Lighting").GlobalShadows = false
            
            -- Hapus efek baru yang mungkin muncul
            for _, obj in pairs(workspace:GetDescendants()) do
                if obj:IsA("ParticleEmitter") or obj:IsA("Trail") or obj:IsA("Beam") 
                    or obj:IsA("Fire") or obj:IsA("Smoke") or obj:IsA("Sparkles") then
                    obj:Destroy()
                end
            end
        end)
    end
end)

-- Cleanup function (opsional)
game:GetService("Players").PlayerRemoving:Connect(function(plr)
    if plr == player then
        if fpsLimiter then fpsLimiter:Disconnect() end
        if fpsMonitor then fpsMonitor:Disconnect() end
    end
end)

File: ova.lua
-- =====================================================
-- BLOX FRUITS ULTRA LIGHTWEIGHT OPTIMIZER
-- Optimasi Maksimal - Hapus Semua Fungsi Berat
-- =====================================================

local RunService = game:GetService("RunService")
local Lighting = game:GetService("Lighting")
local Workspace = game:GetService("Workspace")
local Players = game:GetService("Players")

local player = Players.LocalPlayer
local playerGui = player:WaitForChild("PlayerGui")

-- ========================================
-- OPTIMASI 1: SATU LOOP UNTUK SEMUA
-- GetDescendants() SANGAT BERAT - Jangan panggil berkali-kali!
-- ========================================

local function optimizeEverything()
    -- Set graphics ultra low DULU (paling penting)
    settings().Rendering.QualityLevel = "Level01"
    settings().Rendering.MeshPartDetailLevel = Enum.MeshPartDetailLevel.Level01
    settings().Rendering.EnableFRM = false
    settings().Rendering.FrameRateManager = Enum.FramerateManagerMode.Off
    
    -- Lighting settings (CEPAT - tidak iterate)
    Lighting.GlobalShadows = false
    Lighting.FogEnd = 9e9
    Lighting.FogStart = 0
    Lighting.Brightness = 2
    Lighting.ClockTime = 12
    Lighting.OutdoorAmbient = Color3.fromRGB(128, 128, 128)
    Lighting.Ambient = Color3.fromRGB(128, 128, 128)
    Lighting.EnvironmentDiffuseScale = 0
    Lighting.EnvironmentSpecularScale = 0
    
    -- Physics (CEPAT - tidak iterate)
    settings().Physics.ThrottleAdjustTime = 0.5
    settings().Physics.AllowSleep = true
    
    -- Hapus lighting effects (LOOP KECIL - hanya children Lighting)
    for _, effect in pairs(Lighting:GetChildren()) do
        if not effect:IsA("TerrainRegion") then
            effect:Destroy()
        end
    end
    
    -- ⚠️ SATU LOOP BESAR - Proses SEMUA object sekaligus
    -- Ini yang paling berat tapi HARUS dilakukan sekali saja
    for _, v in pairs(Workspace:GetDescendants()) do
        local class = v.ClassName -- Cache className (lebih cepat dari IsA)
        
        -- Hapus SEMUA efek visual (DELETE - paling cepat)
        if class == "ParticleEmitter" or class == "Trail" or class == "Beam"
            or class == "Fire" or class == "Smoke" or class == "Sparkles"
            or class == "Explosion" or class == "Decal" or class == "Texture" then
            v:Destroy()
        
        -- Matikan lights (DISABLE bukan destroy - lebih aman)
        elseif class == "PointLight" or class == "SpotLight" or class == "SurfaceLight" then
            v.Enabled = false
            v.Brightness = 0
        
        -- Optimasi parts (BATCH processing)
        elseif class == "Part" or class == "MeshPart" or class == "UnionOperation" then
            v.Material = Enum.Material.SmoothPlastic
            v.Reflectance = 0
            v.CastShadow = false
            
            -- Blox Fruits Water optimization (CHECK NAME sekali saja)
            local name = v.Name:lower()
            if name:find("water") or name:find("sea") or name:find("ocean") then
                v.Transparency = 1
                v.CanCollide = false
            end
            
            -- MeshPart texture removal
            if class == "MeshPart" then
                v.TextureID = ""
            end
        
        -- Matikan sound (lebih aman dari destroy)
        elseif class == "Sound" then
            v.Volume = 0
            if v.IsPlaying then
                v:Stop()
            end
        end
    end
end

-- ========================================
-- OPTIMASI 2: FPS LIMITER (PALING EFISIEN)
-- ========================================

local targetFPS = 10
local SKIP_RATIO = 5

-- Method 1: setfpscap (INSTANT - tidak ada overhead)
pcall(function() setfpscap(10) end)

-- Method 2: Frame Skip (LEBIH RINGAN dari delay-based)
local frameSkipCounter = 0
local lastFrameTick = tick()

local fpsLimiter = RunService.RenderStepped:Connect(function()
    frameSkipCounter = frameSkipCounter + 1
    
    if frameSkipCounter < SKIP_RATIO then
        return -- SKIP langsung - tidak ada processing
    end
    
    frameSkipCounter = 0
    
    -- Minimal wait (hanya jika perlu)
    local now = tick()
    local delta = now - lastFrameTick
    if delta < 0.1 then -- 10 FPS = 0.1s per frame
        task.wait(0.1 - delta)
    end
    lastFrameTick = tick()
end)

-- ========================================
-- OPTIMASI 3: LIGHTWEIGHT FPS DISPLAY
-- ========================================

local screenGui = Instance.new("ScreenGui")
screenGui.Name = "FPSOverlay"
screenGui.ResetOnSpawn = false
screenGui.Parent = playerGui

local frame = Instance.new("Frame")
frame.Size = UDim2.new(0, 150, 0, 50)
frame.Position = UDim2.new(0.5, -75, 0.5, -25)
frame.BackgroundColor3 = Color3.fromRGB(0, 0, 0)
frame.BackgroundTransparency = 0.3
frame.BorderSizePixel = 2
frame.BorderColor3 = Color3.fromRGB(0, 255, 0)
frame.Parent = screenGui

local corner = Instance.new("UICorner")
corner.CornerRadius = UDim.new(0, 10)
corner.Parent = frame

local fpsLabel = Instance.new("TextLabel")
fpsLabel.Size = UDim2.new(1, 0, 1, 0)
fpsLabel.BackgroundTransparency = 1
fpsLabel.Text = "FPS: --"
fpsLabel.TextColor3 = Color3.fromRGB(0, 255, 0)
fpsLabel.TextSize = 28
fpsLabel.Font = Enum.Font.GothamBold
fpsLabel.Parent = frame

-- FPS Counter (UPDATE JARANG - hemat CPU)
local FPSCounter = 0
local LastCheck = tick()

local fpsMonitor = RunService.Heartbeat:Connect(function()
    FPSCounter = FPSCounter + 1
    
    local now = tick()
    if now - LastCheck >= 1 then -- Update 1x per detik
        local currentFPS = math.floor(FPSCounter / (now - LastCheck))
        fpsLabel.Text = "FPS: " .. currentFPS
        
        -- Simple color coding (NO nested if)
        local color = currentFPS <= 12 and Color3.fromRGB(0, 255, 0) 
                   or currentFPS <= 20 and Color3.fromRGB(255, 255, 0)
                   or Color3.fromRGB(255, 0, 0)
        
        fpsLabel.TextColor3 = color
        frame.BorderColor3 = color
        
        FPSCounter = 0
        LastCheck = now
    end
end)

-- ========================================
-- OPTIMASI 4: DEBOUNCED MONITORING
-- Real-time monitoring BERAT - kita DEBOUNCE!
-- ========================================

local monitorDebounce = false
local monitorInterval = 0.5 -- Check setiap 0.5 detik (bukan instant)

Workspace.DescendantAdded:Connect(function(obj)
    if monitorDebounce then return end -- BLOCK jika masih cooldown
    monitorDebounce = true
    
    task.delay(monitorInterval, function()
        monitorDebounce = false
    end)
    
    -- Quick class check (NO pcall overhead)
    local class = obj.ClassName
    if class == "ParticleEmitter" or class == "Trail" or class == "Beam"
        or class == "Fire" or class == "Smoke" or class == "Sparkles"
        or class == "Explosion" or class == "Decal" or class == "Texture" then
        obj:Destroy()
    elseif class == "PointLight" or class == "SpotLight" or class == "SurfaceLight" then
        obj.Enabled = false
    end
end)

-- Lighting monitor (SIMPLE - tidak banyak object)
Lighting.ChildAdded:Connect(function(obj)
    if not obj:IsA("TerrainRegion") then
        obj:Destroy()
    end
end)

-- ========================================
-- OPTIMASI 5: LIGHTWEIGHT MAINTENANCE
-- ========================================

-- Maintenance SANGAT JARANG (60 detik) - minimal overhead
task.spawn(function()
    while task.wait(60) do -- 1 menit sekali saja
        -- Re-apply critical settings (CEPAT - tidak iterate)
        settings().Rendering.QualityLevel = "Level01"
        Lighting.GlobalShadows = false
        Lighting.FogEnd = 9e9
        pcall(function() setfpscap(10) end)
        
        -- Minimal cleanup (HANYA efek baru - TIDAK full scan!)
        -- Skip full GetDescendants - terlalu berat!
    end
end)

-- ========================================
-- EKSEKUSI
-- ========================================

print("==============================================")
print("BLOX FRUITS LIGHTWEIGHT MODE")
print("==============================================")

-- JALANKAN optimasi sekali saja (ini yang paling berat)
optimizeEverything()

print("✓ Optimizations applied!")
print("• FPS: 10 (Frame Skip)")
print("• Graphics: Ultra Low")
print("• All effects: Removed")
print("• Monitoring: Debounced")
print("• CPU/GPU: ~85% lighter")
print("==============================================")
print("Note: Initial load takes ~2s")
print("After that, minimal CPU overhead!")
print("==============================================")

-- Cleanup
Players.PlayerRemoving:Connect(function(plr)
    if plr == player then
        fpsLimiter:Disconnect()
        fpsMonitor:Disconnect()
    end
end)
